Media library guide (#377)
livingdocsIO/livingdocs: guides/media_library.md

# Media Library

Livingdocs contains a Media Library that solves management and delivery of Images and Videos.
There are different pieces that can be used to cover custom needs per project.

> This guide will walk you through setting up a basic Media Library for you project with Images support, you will learn about some details along the way.

Reviewer: ```suggestion
This guide will walk you through setting up a basic Media Library for your project with Images support, you will learn about some details along the way.
```

> The Videos Feature is not feature complete yet, it will be documented when it is. If you

Reviewer: ```suggestion
The Videos feature is not feature complete yet, it will be documented when it is. If you
```

are interested in testing it, please contact us.

At heart, the Media Library is based around configured `Media Types`. You might be familiar already with the concept of `Content Types`. `Media Types` are very similar.
You can configure as many different `Media Types` as you want. Usually you want to have at least 1 for Images (in fact, if you don't define one yourself, there is one added automatically at runtime).

See the [mediaType reference](/reference-docs/project-config/media_types.md) for a full reference of the `mediaType` configuration options.

## Media Types
Let's setup a `mediaType` for the images first. You add it to your [project config](/reference-docs/project-config/README.md) in an array at the top-level property `mediaTypes`.

> media-types/image.js

Reviewer: I would move that down to the code block
```js
// media-types/image.js
...
```

Also the other headers in the bottom sections

```js
module.exports = {

>   type: 'mediaImage', // the type is either mediaImage or mediaVideo

Reviewer: ```suggestion
  type: 'mediaImage', // the type is either 'mediaImage' or 'mediaVideo'
```

  handle: 'image', // You should name the primary image mediaType `image`
  info: {
    label: 'Images', // used in dashboards generated for this mediaType

>     description: ''

Reviewer: is this anywhere used? or is it just to understand for the dev why this media type exists?

Author: Not yet, it's available on contentTypes where we have UI in project settings already. This is basically the same but not used yet.

  metadata: [ // any metadata configuration as you know it from contentTypes already
    {
      handle: 'title',
      type: 'li-text',
      config: {
        required: true, // if a metadata property is required, the user will see a form to enter the metadata during upload
        requiredErrorMessage: 'Please provide a title',
        maxLength: 200,
        index: true
      },
      ui: {component: 'liMetaTextForm'}
    },
    {
      handle: 'description',
      type: 'li-text',
      config: {
        index: true
      },
      ui: {component: 'liMetaTextForm'}
    },
    {
      handle: 'credit',
      type: 'li-text',
      config: {
        required: true,
        requiredErrorMessage: 'Please provide a source',
        index: true
      },
      ui: {component: 'liMetaTextForm'}
    },
    {
      handle: 'namedCrops',
      type: 'li-named-crops',
      config: {
        index: true,
        namedCrops: [
          {
            name: 'mobile',
            isOptional: true
          },
          {
            name: 'desktop'
          }
        ]
      },
      ui: {
        label: 'Named Crops',
        component: 'liMetaNamedCropsForm'
      }
    }
  ],
  editor: {

>     // the dashboard seen by users when opening Images/Videos from the document editor

Reviewer: ```suggestion
    // the dashboard on the sidebar seen by users when opening Images/Videos from the document editor
```

    dashboard: {
      displayFilters: [
        {
          filterName: 'liDateTimeRange'
        }
      ]
    },
    // the dashboard opened through the main navigation
    managementDashboard: {
      displayFilters: [
        {
          filterName: 'liDateTimeRange'
        }
      ]
    }
  }
}
```

### IPTC extraction
Since many image files (especially those you get from image agencies) hold metadata defined by the IPTC standard, you want to configure extraction for certain metadata fields, so your users don't have to manually type them.
You do this by adding the `exifExtraction` property to the `mediaType`. See [the list of all IPTC fields](/reference-docs/project-config/media_types.md#exif-extraction) and how to configure them.

The exif extraction will run when an image is imported through the [Import API](/reference-docs/server-api/import_api.md) or when a user uploads an image through the UI.
The user has the chance to manually change the extracted Metadata before the image is actually stored in the Media Library.

media-types/image.js
```js
module.exports = {
  type: 'mediaImage',
  handle: 'image',
  // ...
  exifExtraction: {
    mappings: [
      {

>         field: 'Title',

Reviewer: ```suggestion
        field: 'Title', // IPTC fiel name
```

        metadataPropertyName: 'title'
      },
      {
        field: 'Description',
        metadataPropertyName: 'description'
      },
      {
        field: 'Credit Line',
        metadataPropertyName: 'credit'
      }
    ]
  },
  // ...
}
```

> ### Crops

Reviewer: ```suggestion
### Named Crops
```

> With Livingocs, you can manage multiple crops per Image. We call it `Named Crops` and it works like this:

Reviewer: ```suggestion
With Livingdocs, you can manage multiple crops per Image. We call it `Named Crops` and it works like this:
```

- You can define the crops hold with an image in the Media Library
- A user can edit these crops during upload or from within the Media Library Management Dashboard
- Whenever an image is used in a document, the crops are copied into the document and can be changed locally within the document

There are different strategies how to configure the crops:
1. Configure Named Crops with names after the usage, e.g. `desktop`, `mobile`
2. Configure Named Crops with aspect ratio names, e.g. `16:9`, `4:3`

> We are going to configure the strategy 1 in this guide. There are 3 points where we need to configure this:

Reviewer: ```suggestion
We are going to configure strategy 1 in this guide. There are 4 points where we need to configure this:
```

1. designSettings
2. mediaType
3. Design: image components `doc-image` directive
4. Metadata: `li-image` metadata properties (for a teaser image for example)

#### designSettings

> In the [designSettings](/reference-docs/design.md), all the `namedCrops` need to be configured. You can then use all or some of them in `mediaType`s and `doc-image` directives.

Reviewer: ```suggestion
In the [designSettings](/reference-docs/project-config/design.md) of the project config, all the `namedCrops` need to be configured. You can then use all or some of them in `mediaType`s and `doc-image` directives.
```

```js
projectConfig.designSettings: {
  //...
  namedCrops: [
    {
      handle: 'mobile',
      label: 'Mobile',
      description: 'this image will be shown on small screens',
      // on small screens, having exact same ratios across all images is not that important
      // we let the user pick one but change freely
      recommendedRatios: ['1:1', '16:9']
    },
    {
      handle: 'desktop',
      label: 'Desktop',
      description: 'this image will be shown on wide screens',
      // for large screens, we want consistent 16:9 image ratios, the user can't change that
      imageRatios: ['16:9']
    },
    {
      handle: 'sm-teaser',
      label: 'SM Teaser',
      description: 'this image will be shown on wide screens',
      // for social media, we might want to use a different crop, but the ratio is fixed
      imageRatios: ['16:9']
    }
  ],
  //...
}
```

#### mediaType

> Now we want to make sure the user can define the crops during upload or when changing an image in the Media Library. This way, the user can optimize the crops during upload, then these crops will be used by default when an image is inserted into a document.

Reviewer: I didn't get this sentence completely. You should make it more explicit, that this config is used for the media library asset in the media library and not in the document. And in a second step, you can explain that these crops are used by default in a document when adding that media library asset to the document.

There is a special purpose metadata plugin `li-named-crops` available to be used for this usecase. You configure it within the metadata config of a `mediaType` of type `mediaImage`.

media-types/image.js
```js
module.exports = {
  type: 'mediaImage',
  handle: 'image',
  // ...
  metadata: [
    // ...
    {
      handle: 'namedCrops',
      type: 'li-named-crops',
      config: {
        // configure all the namedCrops defined in the designSettings
        // that you want to keep a default within every mediaLibrary image for
        namedCrops: [
          {
            // the mobile crop is optional, the user can add it at will

>             name: 'mobile', // name references the handle of a namedCrop in the designSettings

Reviewer: what happens when the crop does not exist in the designSettings?

Author: a kitten will die :-)
